docs/java-rest/high-level/migration.asciidoc

```diff
@@ -0,0 +1,137 @@
+
+This dependency must be removed and replaced by a new dependency on the
+high-level client. The <<java-rest-high-usage,Getting Started>> page shows
+ typical configurations for Maven and Gradle and presents the
+ <<java-rest-high-usage-dependencies, dependencies>> brought by the
+ high-level client.
+
+==== Changing the client's initialization code
+
+The `TransportClient` is typically initialized with something like:
+[source,java]
+--------------------------------------------------
+Settings settings = Settings.builder()
+        .put("cluster.name", "prod").build();
+
+TransportClient client = new PreBuiltTransportClient(settings)
+        .addTransportAddress(new TransportAddress(InetAddress.getByName("host"), 9300));
+--------------------------------------------------
+
+The initialization of a `RestHighLevelClient` is different. It first requires the initialization
+of a <<java-rest-low-usage-initialization,low-level client>>:
+
+[source,java]
+--------------------------------------------------
+RestClient lowLevelRestClient = RestClient.builder(
+                new HttpHost("host", 9200, "http")).build();
+--------------------------------------------------
+
+Which is then passed to the constructor of the `RestHighLevelClient`:
+
+[source,java]
+--------------------------------------------------
+RestHighLevelClient client =
+    new RestHighLevelClient(lowLevelRestClient);
+--------------------------------------------------
+
+Both `RestClient` and `RestHighLevelClient` are thread safe. They are
+ typically instantiated by the application at startup time or when the
+ first request is executed, and must both be closed when they are not
+ needed anymore or when the application is stopped.
+
+The `RestClient` provides various <<java-rest-low-usage-initialization,
+ initialization options>> that can be used to tune the way it executes
+ the requests.
+
+Once the `RestHighLevelClient` is initialized, it can then be used to
+execute any of the <<java-rest-high-supported-apis,supported APIs>>.
+
+WARNING: The high-level client does not verify the name of the cluster
+it connects to. This verification can be made after the initialization
+of the client by checking the <<java-rest-high-main, cluster's information>>
+
+==== Changing the application's code
+
+The `RestHighLevelClient` does not provide the same methods than
+the `TransportClient` to execute the requests.
+
+More important, the high-level client:
+
+- does not provide indices or cluster management APIs. Management
+operations must be executed by external scripts or using the low-level
+client.
+- does not provide request builders like the `TransportClient`. The
+ legacy methods like `client.prepareIndex()` must be changed to use
+ request constructors like `new IndexRequest()` to create requests
+ instances. The requests are then executed using synchronous or
+ asynchronous dedicated methods like `client.index()` or `client.indexAsync()`.
```

Comment:
> Do we want to give people the hint to keep the transport dependency, create a dummy client, and do client.prepareIndex(index, type, id).request() to retrieve the request, which can be reused with the hl client? I am not sure as it is a temporary solution but it can save time in the beginning.

I think it makes sense as an automated migration mechanism but I don't particularly like the idea of telling people to do it. Maybe something like:

While you are incrementally migrating your application and you have both the transport client and the high level client available you can always get the Request object from the Builder object by calling `Builder.request()`. We do not advise continuing to depend on the builders in the long run but it should be possible to use them during the transition from the transport client to the high level rest client.


I dunno. Now that I write that it just sounds like a bunch of words.